### PR TITLE
add register route to create new users, add error handling for all cases

### PR DESCRIPTION
**Pull request by kevintrinh388**
Added register route, added error handling to handle all cases/scenarios

### CODE REVIEW

**Comment by kevintrinh388** [comment]
[Create register route](https://trello.com/c/lJZBCHy2/7-create-register-route)

File: app.py [hunk 1]
```diff
@@ -37,6 +39,41 @@ def users():
 
     return jsonify(res)
 
+@app.route("/user", methods=["POST"])
+def register():
+    data = request.get_json(force=True)
+    email = data["email"]
+    password = data["password"]
+    if len(email)==0 and len(password)==0:
+        return "Email and password are empty",400
```
Comment (kphan32) [inline app.py:48]: nit: formatting

File: app.py [hunk 2]
```diff
@@ -37,6 +39,41 @@ def users():
 
     return jsonify(res)
 
+@app.route("/user", methods=["POST"])
+def register():
+    data = request.get_json(force=True)
+    email = data["email"]
+    password = data["password"]
+    if len(email)==0 and len(password)==0:
```
Comment (kphan32) [inline app.py:47]: are empty values actually passed as empty strings? or are they `None`/do not exist within the dictionary?

File: app.py [hunk 3]
```diff
@@ -37,6 +39,41 @@ def users():
...
+
+    try:
+      # validate and get info
+        validate_email(email)
+    except EmailNotValidError:
+        # email is not valid
+        return "invalid email", 400
+
+    if len(password)==0:
+        return jsonify({"error":"invalid password"}), 400
```
Comment (kphan32) [inline app.py:58]: this already gets checked

File: app.py (hunk 1 above)
Comment (kphan32) [inline app.py:48]: i would make the if an `or` and rewrite the error to be `must provide email and password`

Comment (kphan32) [reply app.py:48]: also, make this into JSON like below

File: app.py [hunk 4]
```diff
@@ -37,6 +39,41 @@ def users():
...
+
+    user = User.query.filter_by(email=email).first()
+    if user:
+        return "User with email already exists", 409
```
Comment (kphan32) [inline app.py:62]: ideally this check would be done by the database (ie when creating, it raises an error about validating a unique-ness index)

File: app.py [hunk 5]
```diff
@@ -37,6 +39,41 @@ def users():
...
+
+    hashed_pass = hashlib.sha256(str.encode(password)).hexdigest()
```
Comment (kphan32) [inline app.py:64]: use bcrypt to make your life easier : )

**Review by kphan32 — CHANGES_REQUESTED** [review]

File: app.py (hunk 5 above)
Comment (kevintrinh388) [reply app.py]: done

File: app.py (hunk 3 above)
Comment (kevintrinh388) [reply app.py:58]: done

File: app.py (hunk 1 above)
Comment (kevintrinh388) [reply app.py:48]: done

Comment (kevintrinh388) [reply app.py:48]: done

File: app.py (hunk 2 above)
Comment (kevintrinh388) [reply app.py:47]: depends on what the frontend could possibly send, I have now covered both scenarios
1. the request has no email field OR no password field
2. the request's email OR password field is null OR empty string
done

File: app.py (hunk 4 above)
Comment (kevintrinh388) [reply app.py]: done

File: app.py [hunk 6]
```diff
@@ -36,3 +39,38 @@ def users():
...
+    new_user = User(
+        email=email,
+        password=hashed_pass
+    )
+
+    try:
+        db.session.add(new_user)
+        db.session.commit()
+    except SQLAlchemyError:
+        return jsonify({"error":"User with email already exists"}), 409
```
Comment (kphan32) [inline app.py:74]: can we verify that this is an actual unique key restraint error? it looks like it could be any error and we'd respond with a 409

**Review by kphan32 — CHANGES_REQUESTED** [review]

File: app.py (hunk 6 above)
Comment (kphan32) [reply app.py:74]: (in case of an unrelated DB failure, return 500 and log the error)

Comment (kphan32) [reply app.py:74]: (or maybe even just re-raise the error? not sure how flask will handle that though)

**Review by kphan32 — APPROVED** [review]
lgtm :+1:

**Pull request merged**